snuba/clickhouse/formatter.py

```diff
@@ -0,0 +1,95 @@
+from typing import Optional, Sequence
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Literal,
+)
+from snuba.query.parsing import ParsingContext
+from snuba.util import escape_alias, escape_col, escape_string
+
+# Tokens used when formatting. Defining them as costant
+# will make it easy (if/when needed) to make some changes
+# to the layout of the output.
+NULL = "NULL"
+COMMA = ","
+SPACE = " "
+TRUE = "true"
+FALSE = "false"
+QUOTE = "'"
+OPEN_BRACKET = "("
+CLOSED_BRACKET = ")"
+DOT = "."
+
+
+class ClickhouseExpressionFormatter(ExpressionVisitor[str]):
+    """
+    This Visitor implementation is able to format one expression in the Snuba
+    Query for Clickhouse.
+
+    The only state mainained is the PrasingContext, which allows us to resolve
+    aliases and can be reused when formatting multiple expressions.
+
+    When passing an instance of this class to the accept method of
+    the visited expression, the return value is the formatted string.
+
+    This Formatter produces a properly escaped string. The result should never
+    be further escaped. This should be the only place where expression
+    escaping happens as it is done by each method that formats a specific
+    type of expression.
+    """
+
+    def __init__(self, parsing_context: Optional[ParsingContext] = None) -> None:
+        self.__parsing_context = parsing_context or ParsingContext()
+
+    def __alias(self, formatted_exp: str, alias: Optional[str]) -> str:
+        if not alias:
+            return formatted_exp
+        elif self.__parsing_context.is_alias_present(alias):
+            ret = escape_alias(alias)
```

Comment:
For now I relied on the existing escaping functions instead of providing a new implementation here. I don't think adding a second implementation (which would be simpler) gives us any advantage now.

Reply:
Makes sense, thanks for pointing that out, the comment below is helpful.